src/core/shape_inference/include/gru_sequence_shape_inference.hpp

```diff
@@ -0,0 +1,178 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#pragma once
+#include <openvino/core/validation_util.hpp>
+#include <openvino/op/gru_sequence.hpp>
+
+#include "utils.hpp"
+
+namespace ov {
+namespace op {
+namespace rnn_seq {
+template <class OpType, class ShapeType>
+void validate_inputs_rank(const OpType* op,
+                          const std::vector<ShapeType>& input_shapes,
+                          const std::vector<Rank>& expected_ranks) {
+    NODE_VALIDATION_CHECK(op, input_shapes.size() >= expected_ranks.size(), "Can't validate inputs rank.");
+    for (auto i = 0; i < expected_ranks.size(); ++i) {
+        NODE_VALIDATION_CHECK(op,
+                              input_shapes[i].rank().compatible(expected_ranks[i]),
+                              "Shape rank of input at ",
+                              i,
+                              " is incompatible. Expected rank: ",
+                              expected_ranks[i],
+                              ", actual shape: ",
+                              input_shapes[i],
+                              ".");
+    }
+}
+
+// Output shapes layout:
+// output_shapes[0]: [batch_size, num_directions, seq_length, hidden_size] // Rank always 4
+// output_shapes[1]: [batch_size, num_directions, hidden_size] // Rank always 3
+template <class OpType, class ShapeType>
+void gru_shape_infer(const OpType* op,
+                     const std::vector<ShapeType>& input_shapes,
+                     std::vector<ShapeType>& output_shapes) {
+    NODE_VALIDATION_CHECK(op,
+                          input_shapes.size() >= 6 && output_shapes.size() == 2,
+                          "Incorrect number of shapes has been provided.");
+
+    auto& y_out_shape = output_shapes[0];
+    auto& ho_out_shape = output_shapes[1];
+    y_out_shape.resize(4);   // Rank always 4
+    ho_out_shape.resize(3);  // Rank always 3
+
+    rnn_seq::validate_inputs_rank(op, input_shapes, {3, 3, 1, 3, 3, 2});
```

Comment:
1. Sounds this validate_inputs_rank might be invoked before rank().is_static() check. Could be problem if rank is dynamic?

2. We should make shape_infer return as quick as possible, as it is frequently invoked not only in compiling but also in inferencing by both IE and plugins. So once the PS is of dynamic rank and we cannot propagate shapes, we'd better return immediately.

Reply:
Thank you for your comments @ceciliapeng2011.

1. This validation works with dynamic rank, and accepts it as "valid" rank.
Dynamic rank is compatible with any "static", so rank().compatible is flexible:
                              `input_shapes[i].rank().compatible(expected_ranks[i])`

Such case is tested with type_prop tests:
`gru_sequence_all_inputs_dynamic_rank`,
`augru_sequence_all_inputs_dynamic_rank`

2. I mentioned in the description that this PR introduces an improvement - output rank is always static.
According to the spec and implementation:
https://github.com/openvinotoolkit/openvino/blob/a74ffa15bc712b2edbc265cba9b0029cc79f738d/src/core/shape_inference/include/gru_sequence_shape_inference.hpp#L31-L33
Input shape information is not needed to set the ranks for the outputs.
Also if one input has dynamic rank, it doesn't mean we can't propagate some information about dimensions from other inputs.

About the case when `shape_infer` is called by plugins, as I understand for CPU the `shape_infer` is called with `StaticShape` as the template type, so it's not possible that the dynamic rank case will be executed for plugin anyway.